module: expose `resolveLoadAndCache` API (#55756)
nodejs/node: doc/api/module.md

This feature requires `--allow-worker` if used with the [Permission Model][].

### `module.resolveLoadAndCache(specifier[, parentURL[, importAttributes[, conditions]]])`

<!-- YAML
added: REPLACEME
-->

> Stability: 1 - Experimental

* `specifier` {string|URL} Module to resolve and load; this should be
  the same string that would be passed to `import()`, except that if it is
  relative, it is resolved relative to `parentURL`.
* `parentURL` {string|URL|undefined} If you want to resolve `specifier` relative to a base
  URL, such as `import.meta.url`, you can pass that URL here. If not provided,

>   the `resolve` step will be skipped. **Default:** {undefined}.

Reviewer: Not a fan of implicitly skip `resolve` when `parentURL` is `undefined`. This also contradicts with the API name `resolve, load, and cache`.

We have several conventions in APIs about how to resolve the `parentURL`:
- `new Worker(filename)` resolves with base to `process.cwd()`,
- `module.register(specifier[, parentURL])` resolves with base to `data:` if `parentURL` is not set.

I think as part of module loader API, it is better to be explicit, e.g. throw if `parentURL` is missing.

Author: > * `module.register(specifier[, parentURL])` resolves with base to `data:` if `parentURL` is not set.
> 
> I think as part of module loader API, it is better to be explicit, e.g. throw if `parentURL` is missing.

Note that this way of doing things is consistent with `module.register`, i.e. it will throw if `specifier` is not a full URL and `parentURL` is missing (because `load` only accepts full URLs).

> Not a fan of implicitly skip `resolve` when `parentURL` is `undefined`. This also contradicts with the API name `resolve, load, and cache`.

The name is not great, I'm certainly fine changing it to something else. IMO being able to skip `resolve` is important as it allows to get results guaranteed to be true for the lifetime of the process (at least until we expose a way to clear the load cache), I'm open to suggestions if you think we can make a clearer API

Reviewer: What I'm concerned about is that using a full `specifier` URL and an implicit `parentURL` to indicate skipping `resolve` is not alike with `import`:

```js
import 'test:./foobar'
```

With import statements or import calls, a resolve hook will be invoked with a full `specifier` URL and implicit `parentURL`.

```js
export function resolve(specifier, context, next) {
  if (specifier.startsWith('test:')) {
    // Translate custom protocol and use the default loader.
    return next(specifier.slice(5), context);
  }
  return next(specifier, context);
}
```

Author: We cannot make it work like `import`, unless we define it on `import.meta`, but I get what you're saying. What do you think about going back to using `data:` as default `parentURL`, and expose some symbol to explicitly skip the resolve call if needed? 
```js
// does not go through `resolve` hook:
resolveLoadAndCache('full://module/url', module.SKIP_RESOLVE_CALL);
// or maybe
resolveLoadAndCache('full://module/url', Symbol.for('nodejs.skipResolveCall'));

// does go through `resolve` hook with `data:` as `parentURL`:
resolveLoadAndCache('full://module/url');
```

Reviewer: Yeah, I agree that this API does not infer `parentURL` implicitly to the caller module (it is not practical anyway). I think a possible alternative to Symbol based API is separating the motivation of invoking each of the loader hook into individual APIs:

```typescript
interface ModuleLoaderAPI {
  resolve(specifier, parentURL, attributes): Promise<{ url, format }>;
  load(url, attributes): Promise<{ format, source }>;

  resolveAndLoad(specifier, parentURL, attributes): Promise<{ url, format, source }>;
}
```

Register a module that exports [hooks][] that customize Node.js module
resolution and loading behavior. See [Customization hooks][].

### `module.resolveLoadAndCache(specifier[, parentURL[, importAttributes[, conditions]]])`

<!-- YAML
added: REPLACEME
-->

> Stability: 1 - Experimental

> * `specifier` {string|URL} Customization hooks to be registered; this should be

Author: ```suggestion
* `specifier` {string|URL} Module to load; this should be
```

  the same string that would be passed to `import()`, except that if it is
  relative, it is resolved relative to `parentURL`.
* `parentURL` {string|URL} If you want to resolve `specifier` relative to a base
  URL, such as `import.meta.url`, you can pass that URL here. **Default:**
  `'data:'`.
* `importAttributes` {object}
* `conditions` {Array}
* Returns: {Promise} fulfills with an object with the following properties:
  * `url` {string} The absolute URL for that module
  * `format` {string} The format this module will be parsed as.
  * `source` {null|TypedArray}

> This API tells you how a specific URL will be loaded by the ECMAScript loader if

Reviewer: FWIW, there is no "ECMAScript loader" (yet).

```suggestion
This API tells you how a specific URL will be loaded by the module loader if
```

it was imported from the `parentURL` in the current process. If the module was
already imported before `resolveLoadAndCache` is called, the cached version is
returned; if not, it will populate the cache so future calls to

> `resolveLoadAndCache` or `import` do re-do the work.

Reviewer 2: Shouldn't this be:

```suggestion
`resolveLoadAndCache` or `import` do not re-do the work.
```
